### ISSUE TITLE
Proposal: Initializer is unnecessary

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn-analyzers#3036]
Fields in C# typically do not need to be initialized to their default values. This is especially true for `static readonly` fields, such as `Guid.Empty`. An analyzer could report cases where a field is unnecessarily initialized to a default value, and offer a code fix to remove the initialization.

**Comment by sharwell** [comment]
@stephentoub It would help this proposal if you could identify the cases where a field initializer for a default value is not redundant, and therefore should not be reported by this analyzer.

**Comment by stephentoub** [comment]
> It would help this proposal if you could identify the cases where a field initializer for a default value is not redundant, and therefore should not be reported by this analyzer.

The main one that comes to mind is CS0649, e.g. this generates a warning:
```C#
public readonly struct MyStruct
{
    internal static readonly MyStruct Empty;
}
```
whereas this does not:
```C#
public readonly struct MyStruct
{
    internal static readonly MyStruct Empty = default;
}
```
It'd be problematic if an analyzer warned that you should convert the latter into the former, unless the compiler also removes CS0649.

Another that comes to mind is when a value isn't necessary from the runtime's perspective, and isn't even emitted in IL by the C# compiler, but where the C# compiler still requires it for nullable reference type reasons, e.g. to silence a warning about non-null field being null.
```C#
private Foo _foo = null!
```

**Comment by mavasani** [comment]
@stephentoub I believe you already implemented this analyzer, correct?

**Comment by stephentoub** [comment]
> I believe you already implemented this analyzer, correct?

Yes, CA1805.

**Comment by carlreinke** [comment]
Why was this implemented as an analyzer and not as a compiler optimization?  As a compiler optimization, everyone would benefit without code changes.
